File: lib/Models/AuthModels.dart
import 'BaseResponse.dart';

// ==================== SessionUserModel ====================
class SessionUserModel {
  SessionUserModel({
    required this.id,
    required this.username,
    required this.email,
    this.phone,
    this.avatar,
    this.gender,
  });

  final String id;
  final String username;
  final String email;
  final String? phone;
  final String? avatar;
  final String? gender; // backend: MALE/FEMALE

  factory SessionUserModel.fromJson(Map<String, dynamic> json) {
    final id = json['id'] as String?;
    final username = json['username'] as String?;
    final email = json['email'] as String?;
    if (id == null || username == null || email == null) {
      throw Exception('Thiếu thông tin người dùng từ backend (id/username/email).');
    }
    return SessionUserModel(
      id: id,
      username: username,
      email: email,
      phone: json['phone'] as String?,
      avatar: json['avatar'] as String?,
      gender: json['gender'] as String?,
    );
  }
}

// ==================== LoginResult ====================
class LoginResult {
  LoginResult({
    required this.user,
    required this.accessToken,
    required this.refreshToken,
  });

  final SessionUserModel user;
  final String accessToken;
  final String refreshToken;

  factory LoginResult.fromJson(Map<String, dynamic> json) {
    final accessToken = json['accessToken'] as String?;
    final refreshToken = json['refreshToken'] as String?;
    if (accessToken == null || refreshToken == null) {
      throw Exception('Thiếu token đăng nhập từ backend.');
    }
    return LoginResult(
      user: SessionUserModel.fromJson(json['user'] as Map<String, dynamic>),
      accessToken: accessToken,
      refreshToken: refreshToken,
    );
  }
}

// ==================== LoginResponseWrapper ====================
class LoginResponseWrapper extends BaseResponse {
  LoginResult? data;

  LoginResponseWrapper({
    required int statusCode,
    required String message,
    this.data,
  }) : super(statusCode: statusCode, message: message);

  factory LoginResponseWrapper.fromJson(Map<String, dynamic> json) {
    final payload = json['data'];
    return LoginResponseWrapper(
      statusCode: (json['statusCode'] as num).toInt(),
      message: json['message'] as String,
      data: payload is Map<String, dynamic> ? LoginResult.fromJson(payload) : null,
    );
  }
}

// ==================== UserResponseWrapper ====================
class UserResponseWrapper extends BaseResponse {
  SessionUserModel? data;

  UserResponseWrapper({
    required int statusCode,
    required String message,
    this.data,
  }) : super(statusCode: statusCode, message: message);

  factory UserResponseWrapper.fromJson(Map<String, dynamic> json) {
    final payload = json['data'];
    return UserResponseWrapper(
      statusCode: (json['statusCode'] as num).toInt(),
      message: json['message'] as String,
      data: payload is Map<String, dynamic>
          ? SessionUserModel.fromJson(payload)
          : null,
    );
  }
}

File: mobile_restaurant/models/AuthModels.ts
// ==================== BaseResponse ====================
export interface BaseResponse {
  statusCode: number;
  message: string;
  status: boolean;
}

export function parseBaseResponse(json: any): BaseResponse {
  const statusCode = Number(json.statusCode);
  // Backend 'message' can be a string or an array of validation error strings
  let message = '';
  if (typeof json.message === 'string') {
    message = json.message;
  } else if (Array.isArray(json.message)) {
    message = json.message.filter(Boolean).join(' ; ');
  } else if (json.message != null) {
    message = String(json.message);
  }
  return {
    statusCode,
    message,
    status: statusCode >= 200 && statusCode < 300,
  };
}

// ==================== SimpleResponse ====================
export interface SimpleResponse extends BaseResponse {
  data?: any;
}

export function parseSimpleResponse(json: any): SimpleResponse {
  const base = parseBaseResponse(json);
  return { ...base, data: json.data };
}

// ==================== SessionUserModel ====================
export interface SessionUserModel {
  id: string;
  username: string;
  email: string;
  phone?: string;
  avatar?: string;
  gender?: string; // backend: MALE/FEMALE
  role?: string;
  branchId?: string;
  branchName?: string;
}

export function parseSessionUserModel(json: any): SessionUserModel {
  if (!json.id || !json.username || !json.email) {
    throw new Error('Thiếu thông tin người dùng từ backend (id/username/email).');
  }
  return {
    id: json.id,
    username: json.username,
    email: json.email,
    phone: json.phone ?? undefined,
    avatar: json.avatar ?? undefined,
    gender: json.gender ?? undefined,
    role: json.role ?? undefined,
    branchId: json.branchId ?? undefined,
    branchName: json.branchName ?? undefined,
  };
}

// ==================== LoginResult ====================
export interface LoginResult {
  user: SessionUserModel;
  accessToken: string;
  refreshToken: string;
}

export function parseLoginResult(json: any): LoginResult {
  if (!json.accessToken || !json.refreshToken) {
    throw new Error('Thiếu token đăng nhập từ backend.');
  }
  return {
    user: parseSessionUserModel(json.user),
    accessToken: json.accessToken,
    refreshToken: json.refreshToken,
  };
}

// ==================== LoginResponseWrapper ====================
export interface LoginResponseWrapper extends BaseResponse {
  data?: LoginResult;
}

export function parseLoginResponseWrapper(json: any): LoginResponseWrapper {
  const base = parseBaseResponse(json);
  return {
    ...base,
    data: json.data && typeof json.data === 'object' ? parseLoginResult(json.data) : undefined,
  };
}

// ==================== UserResponseWrapper ====================
export interface UserResponseWrapper extends BaseResponse {
  data?: SessionUserModel;
}

export function parseUserResponseWrapper(json: any): UserResponseWrapper {
  const base = parseBaseResponse(json);
  return {
    ...base,
    data: json.data && typeof json.data === 'object' ? parseSessionUserModel(json.data) : undefined,
  };
}
